kernel-test-nohz: add alternative to cgroups v1 and improve boot flags

gratian opened the pull request:
The current implementation assumes cgroups v1 mount points, and a cgroup cpuset directory layout configured for LabVIEW RT. It also does not fully take advantage of all possible 'isolcpus' flags.

Address these issues by providing a non-cgroup_v1 test configuration and adding the required boot flags.

AB #2540629

### Testing

- [x] `bitbake kernel-test-nohz-ptest`
- [x] `opkg install kernel-test-nohz-ptest` from local package feed on a NI Linux RT target
- [x] validated that the test starts correctly when run in a configuration where LabVIEW RT cgroup v1 layout is mounted
- [x] validated that the test starts correctly if no cgroups are mounted

Further performance validation left for the ATS.

gratian committed:
kernel-test-nohz: add isolcpus flags to increase isolation

When the 'isolcpus' boot parameter is used without specifying a flag list
it defaults to just 'domain' isolation.

Increase CPU isolation by explicitly specifying:

  - the 'nohz' flag which offloads the residual 1Hz tick to a work-queue
    (which can be moved to a different CPU);

  - the 'managed_irq' flag which provides isolation from managed
    interrupts;

in addition to the 'domain' flag for the boot parameters of this test.

Signed-off-by: Gratian Crisan <gratian.crisan@ni.com>

gratian committed:
kernel-test-nohz: add alternative to cgroups v1

The current implementation assumes cgroups v1 mount points, and a cgroup
cpuset directory layout configured for LabVIEW RT.

This assumption is no longer true after commit
48f0fe9aea1e ("initscripts-nilrt: add LV RT cgroup version detection to init scripts")
if LabVIEW RT is not installed.

The additional cgroup setup steps are really only needed if the cpuset
cgroups are configured for LV RT.

For the case where the LV RT cgroups are not configured the 'isolcpus' boot
flag already provides the CPU isolation required by this test. The affinity
assignment to the isolated CPU can be handled via an alternate code path
that uses sched_setaffinity() directly.

Natinst-AZDO-ID: 2540629
Signed-off-by: Gratian Crisan <gratian.crisan@ni.com>

chaitu236 commented on recipes-kernel/kernel-tests/kernel-test-nohz.bb:
```diff
@@ -40,7 +40,7 @@ pkg_postinst_ontarget:${PN}-ptest:append() {
 
     if [ $ISOLATED_CPU -gt 0 ]; then
-        echo 'set otherbootargs="${otherbootargs} isolcpus='$ISOLATED_CPU' nohz_full='$ISOLATED_CPU' mitigations=off intel_pstate=disable intel_idle.max_cstate=0 processor.max_cstate=0 nosoftlockup mce=ignore_ce audit=0 tsc=nowatchdog"' > /boot/runmode/no-hz-full-params.cfg
+        echo 'set otherbootargs="${otherbootargs} isolcpus=nohz,domain,managed_irq,'$ISOLATED_CPU' nohz_full='$ISOLATED_CPU' mitigations=off intel_pstate=disable intel_idle.max_cstate=0 processor.max_cstate=0 nosoftlockup mce=ignore_ce audit=0 tsc=nowatchdog"' > /boot/runmode/no-hz-full-params.cfg
```
Would we want to mention these params in [this readme](https://github.com/ni/meta-nilrt/blob/nilrt/master/kirkstone/recipes-kernel/linux/nilrt-nohz/README.nohz#L44C55-L44C55)?
Also, for non-LV based applications [the instructions](https://github.com/ni/meta-nilrt/blob/nilrt/master/kirkstone/recipes-kernel/linux/nilrt-nohz/README.nohz#L56) should be updated right?

gratian replied:
README.nohz updated.

mike-petersen-ni commented on recipes-kernel/kernel-tests/kernel-test-nohz.bb:
```diff
@@ -40,7 +40,7 @@ pkg_postinst_ontarget:${PN}-ptest:append() {
 
     if [ $ISOLATED_CPU -gt 0 ]; then
-        echo 'set otherbootargs="${otherbootargs} isolcpus='$ISOLATED_CPU' nohz_full='$ISOLATED_CPU' mitigations=off intel_pstate=disable intel_idle.max_cstate=0 processor.max_cstate=0 nosoftlockup mce=ignore_ce audit=0 tsc=nowatchdog"' > /boot/runmode/no-hz-full-params.cfg
+        echo 'set otherbootargs="${otherbootargs} isolcpus=nohz,domain,managed_irq,'$ISOLATED_CPU' nohz_full='$ISOLATED_CPU' mitigations=off intel_pstate=disable intel_idle.max_cstate=0 processor.max_cstate=0 nosoftlockup mce=ignore_ce audit=0 tsc=nowatchdog"' > /boot/runmode/no-hz-full-params.cfg
```
Not entirely relevant to this change, but it appears that `isolcpus` has been deprecated since ni/linux@b0d40d2b22fe.

gratian replied:
Yeah, it has been marked deprecated and at some point it may get removed, however:

- It is still present in the latest Linus master (6.7-rc1 at this point)
- It has users that oppose it's removal: https://lore.kernel.org/linux-rt-users/CANV1gkdPJ=gqT2nUuGzN=ytOxHN92opoS1aGEDqy3FmTWTmusw@mail.gmail.com/
- We will notice when it gets removed (via this test)
- I would rather keep this test simple for now and not have to add support for the intermediate cgroup support state we're currently in until LV RT upgrades to v2

gratian committed:
linux-nilrt-nohz: update flags and cgroup info in nohz README

Add additional recommended isolcpu flags.
Also expand on various cgroup options available for CPU isolation
depending on installation setup and programming environment used.

Signed-off-by: Gratian Crisan <gratian.crisan@ni.com>

gratian commented:
Changes since V1:

- updated README.nohz per @chaitu236 suggestion
